types/time.go

```diff
@@ -562,165 +564,444 @@ func ParseDateFormat(format string) []string {
-	var (
-		year, month, day, hour, minute, second int
-		fracStr                                string
-		hhmmss                                 bool
-		err                                    error
-	)
+// The only delimiter recognized between a date and time part and a fractional seconds part is the decimal point.
+func splitDateTime(format string, fsp int, selfDecideFsp bool, isFloat bool) (seps []string, realFsp int, warnMsg string, err error) {
+	format = strings.TrimSpace(format)
+	realFsp = fsp
+	var lastSepChar uint8 = 0
+	var totalPartCount = 0
+	if len(format) == 0 {
+		return
+	}
 
-	seps, fracStr := splitDateTime(str)
-	var truncatedOrIncorrect bool
-	switch len(seps) {
-	case 1:
-		l := len(seps[0])
-		switch l {
-		case 14: // No delimiter.
-			// YYYYMMDDHHMMSS
-			_, err = fmt.Sscanf(seps[0], "%4d%2d%2d%2d%2d%2d", &year, &month, &day, &hour, &minute, &second)
-			hhmmss = true
-		case 12: // YYMMDDHHMMSS
-			_, err = fmt.Sscanf(seps[0], "%2d%2d%2d%2d%2d%2d", &year, &month, &day, &hour, &minute, &second)
-			year = adjustYear(year)
-			hhmmss = true
-		case 11: // YYMMDDHHMMS
-			_, err = fmt.Sscanf(seps[0], "%2d%2d%2d%2d%2d%1d", &year, &month, &day, &hour, &minute, &second)
-			year = adjustYear(year)
-			hhmmss = true
-		case 10: // YYMMDDHHMM
-			_, err = fmt.Sscanf(seps[0], "%2d%2d%2d%2d%2d", &year, &month, &day, &hour, &minute)
-			year = adjustYear(year)
-		case 9: // YYMMDDHHM
-			_, err = fmt.Sscanf(seps[0], "%2d%2d%2d%2d%1d", &year, &month, &day, &hour, &minute)
-			year = adjustYear(year)
-		case 8: // YYYYMMDD
-			_, err = fmt.Sscanf(seps[0], "%4d%2d%2d", &year, &month, &day)
-		case 6, 5:
-			// YYMMDD && YYMMD
-			_, err = fmt.Sscanf(seps[0], "%2d%2d%2d", &year, &month, &day)
-			year = adjustYear(year)
-		default:
-			return ZeroDatetime, errors.Trace(ErrInvalidTimeFormat.GenWithStackByArgs(str))
+	var isFirstPartHasYMD = false
+	var firstSepChar uint8 = 0
+
+	if !unicode.IsNumber(rune(format[0])) {
+		// Date format must start with number.
+		seps = nil
+		return
+	}
+	var start = 0
+	for {
+		if start >= len(format) {
+			return
 		}
-		if l == 5 || l == 6 || l == 8 {
-			// YYMMDD or YYYYMMDD
-			// We must handle float => string => datetime, the difference is that fractional
-			// part of float type is discarded directly, while fractional part of string type
-			// is parsed to HH:MM:SS.
-			if isFloat {
-				// 20170118.123423 => 2017-01-18 00:00:00
-			} else {
-				// '20170118.123423' => 2017-01-18 12:34:23.234
-				switch len(fracStr) {
-				case 0:
-				case 1, 2:
-					_, err = fmt.Sscanf(fracStr, "%2d ", &hour)
-				case 3, 4:
-					_, err = fmt.Sscanf(fracStr, "%2d%2d ", &hour, &minute)
-				default:
-					_, err = fmt.Sscanf(fracStr, "%2d%2d%2d ", &hour, &minute, &second)
-				}
-				truncatedOrIncorrect = err != nil
-			}
+
+		var currentSepChar, currentPart, nextStart, warnMsgRet, errRet = nextTimePart(format, start, totalPartCount, lastSepChar)
```

Comment:
`var currentSepChar, currentPart, nextStart, warnMsgRet, errRet = ` --> `currentSepChar, currentPart, nextStart, warnMsgRet, errRet :=` is better?

Reply:
By the way, I don't think too many return values is good for maintaining, how about define a context struct like:

```
type parseTimeContext struct {
	currentSepChar
	currentPart
	nextStart
	....
}
```